File: server/forge_server.py
#!/usr/bin/env python3
"""
STM32 Forge local build server.

Serves the web IDE and gives it a real toolchain:

  GET  /api/status                 toolchain + targets
  GET  /api/tree                   editable source files
  GET  /api/file?path=...          read a file
  PUT  /api/file?path=...          save a file (body = contents)
  POST /api/build                  {"target": "h743"|"bluepill", "project": "doom"|"blinky"}
                                   -> compiler log, per-region memory usage from
                                      the linker, DOOM zone size, firmware image

Binds to 127.0.0.1 only. Standard library only.

  python3 server/forge_server.py [--port 8732]
  then open http://localhost:8732/
"""
import argparse
import base64
import glob
import json
import os
import re
import shutil
import subprocess
import sys
import threading
from http import HTTPStatus
from http.server import SimpleHTTPRequestHandler, ThreadingHTTPServer
from urllib.parse import parse_qs, urlparse

ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), '..'))
FIRMWARE = os.path.join(ROOT, 'firmware')

TARGETS = {
    'h743': {'name': 'STM32H743IITx', 'core': 'Cortex-M7 @ 400 MHz',
             'flash': 2048 * 1024, 'ram': 1024 * 1024},
    'bluepill': {'name': 'STM32F103C8T6 (Blue Pill)', 'core': 'Cortex-M3 @ 72 MHz',
                 'flash': 64 * 1024, 'ram': 20 * 1024},
}
PROJECTS = ['doom', 'blinky']

# Sources shown in the IDE file tree (vendored libraries are left out).
EDITABLE = [
    'firmware/projects/**/*.[ch]',
    'firmware/targets/**/*.[chs]',
    'firmware/targets/**/*.ld',
    'firmware/Makefile',
    'engine/platform/*.c',
    'engine/doomgeneric/*.[ch]',
]

build_lock = threading.Lock()


def find_toolchain():
    candidates = []
    if os.environ.get('ARM_GCC_PATH'):
        candidates.append(os.environ['ARM_GCC_PATH'])
    on_path = shutil.which('arm-none-eabi-gcc')
    if on_path:
        candidates.append(os.path.dirname(on_path))
    candidates += sorted(glob.glob(os.path.expanduser('~/.local/toolchains/*/bin')), reverse=True)
    candidates += sorted(glob.glob('/opt/*arm*/bin'), reverse=True)
    for d in candidates:
        gcc = os.path.join(d, 'arm-none-eabi-gcc')
        if os.path.isfile(gcc) and os.access(gcc, os.X_OK):
            try:
                ver = subprocess.run([gcc, '--version'], capture_output=True, text=True,
                                     timeout=10).stdout.splitlines()[0]
            except (OSError, subprocess.SubprocessError, IndexError):
                continue
            return d, ver
    return None, None


MEM_LINE = re.compile(r'^\s*(\w+):\s+(\d+(?:\.\d+)?)\s*([KMG]?B)\s+(\d+(?:\.\d+)?)\s*([KMG]?B)\s+')
OVERFLOW = re.compile(r"region [`'](\w+)' overflowed by (\d+) bytes")
UNITS = {'B': 1, 'KB': 1024, 'MB': 1024 ** 2, 'GB': 1024 ** 3}


def parse_memory(log):
    """Parses `ld --print-memory-usage` output into [{region, used, size}]."""
    regions = []
    for line in log.splitlines():
        m = MEM_LINE.match(line)
        if m and m.group(1) != 'Memory':
            regions.append({
                'region': m.group(1),
                'used': int(float(m.group(2)) * UNITS[m.group(3)]),
                'size': int(float(m.group(4)) * UNITS[m.group(5)]),
            })
    overflow = {m.group(1): int(m.group(2)) for m in OVERFLOW.finditer(log)}
    for r in regions:
        r['overflow'] = overflow.get(r['region'], 0)
    return regions


def zone_banks(bin_dir, elf):
    """Sizes of the DOOM zone heap banks from the linker symbols, or None."""
    try:
        out = subprocess.run([os.path.join(bin_dir, 'arm-none-eabi-nm'), elf],
                             capture_output=True, text=True, timeout=30).stdout
    except (OSError, subprocess.SubprocessError):
        return None
    sym = {}
    for line in out.splitlines():
        parts = line.split()
        if len(parts) == 3 and re.fullmatch(r'__zone\d_(start|end)', parts[2]):
            sym[parts[2]] = int(parts[0], 16)
    banks = []
    i = 0
    while '__zone%d_start' % i in sym:
        banks.append(sym['__zone%d_end' % i] - sym['__zone%d_start' % i])
        i += 1
    return banks or None


def build(target, project):
    bin_dir, version = find_toolchain()
    if not bin_dir:
        return {'ok': False, 'log': 'arm-none-eabi-gcc not found. Install the Arm GNU Toolchain '
                'or set ARM_GCC_PATH to its bin directory.\n', 'memory': []}
    cmd = ['make', '-C', FIRMWARE, '-j%d' % (os.cpu_count() or 4),
           'TARGET=' + target, 'PROJECT=' + project, 'ARM_GCC_PATH=' + bin_dir]
    out_dir = os.path.join(FIRMWARE, 'build', '%s-%s' % (target, project))
    with build_lock:
        # Always relink so the linker prints the memory usage report.
        for name in ('firmware.elf', 'firmware.bin', 'firmware.hex'):
            try:
                os.remove(os.path.join(out_dir, name))
            except FileNotFoundError:
                pass
        proc = subprocess.run(cmd, capture_output=True, text=True, timeout=900)
    log = proc.stdout + proc.stderr
    result = {
        'ok': proc.returncode == 0,
        'toolchain': version,
        'command': ' '.join(cmd[:1] + cmd[3:-1]),
        'log': log,
        'memory': parse_memory(log),
    }
    if proc.returncode == 0:
        elf = os.path.join(out_dir, 'firmware.elf')
        with open(os.path.join(out_dir, 'firmware.bin'), 'rb') as f:
            image = f.read()
        result['binary'] = base64.b64encode(image).decode()
        result['binarySize'] = len(image)
        result['zoneBanks'] = zone_banks(bin_dir, elf)
        result['zone'] = sum(result['zoneBanks']) if result['zoneBanks'] else None
    return result


def safe_path(rel):
    """Absolute path for `rel` if it is one of the editable files, else None."""
    path = os.path.realpath(os.path.join(ROOT, rel))
    if not path.startswith(ROOT + os.sep) or os.path.relpath(path, ROOT) not in file_tree():
        return None
    return path


def file_tree():
    files = set()
    for pat in EDITABLE:
        for p in glob.glob(os.path.join(ROOT, pat), recursive=True):
            if os.path.isfile(p):
                files.add(os.path.relpath(p, ROOT))
    return sorted(files)


class Handler(SimpleHTTPRequestHandler):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, directory=ROOT, **kwargs)

    def log_message(self, fmt, *args):
        if '/api/' in (args[0] if args else ''):
            sys.stderr.write('%s\n' % (fmt % args))

    def end_headers(self):
        self.send_header('Cache-Control', 'no-store')
        super().end_headers()

    def send_json(self, obj, status=HTTPStatus.OK):
        data = json.dumps(obj).encode()
        self.send_response(status)
        self.send_header('Content-Type', 'application/json')
        self.send_header('Content-Length', str(len(data)))
        self.end_headers()
        self.wfile.write(data)

    def body(self):
        n = int(self.headers.get('Content-Length', 0))
        return self.rfile.read(n) if n else b''

    def do_GET(self):
        url = urlparse(self.path)
        if url.path == '/api/status':
            bin_dir, version = find_toolchain()
            return self.send_json({'toolchain': version, 'toolchainPath': bin_dir,
                                   'targets': TARGETS, 'projects': PROJECTS})
        if url.path == '/api/tree':
            return self.send_json({'files': file_tree()})
        if url.path == '/api/file':
            path = safe_path(parse_qs(url.query).get('path', [''])[0])
            if not path or not os.path.isfile(path):
                return self.send_json({'error': 'not found'}, HTTPStatus.NOT_FOUND)
            with open(path, encoding='utf-8', errors='replace') as f:
                return self.send_json({'path': os.path.relpath(path, ROOT), 'content': f.read()})
        return super().do_GET()

    def do_PUT(self):
        url = urlparse(self.path)
        if url.path != '/api/file':
            return self.send_json({'error': 'not found'}, HTTPStatus.NOT_FOUND)
        path = safe_path(parse_qs(url.query).get('path', [''])[0])
        if not path or not os.path.isfile(path):
            return self.send_json({'error': 'not an editable file'}, HTTPStatus.FORBIDDEN)
        with open(path, 'wb') as f:
            f.write(self.body())
        return self.send_json({'saved': os.path.relpath(path, ROOT)})

    def do_POST(self):
        if urlparse(self.path).path != '/api/build':
            return self.send_json({'error': 'not found'}, HTTPStatus.NOT_FOUND)
        try:
            req = json.loads(self.body() or b'{}')
        except json.JSONDecodeError:
            return self.send_json({'error': 'bad json'}, HTTPStatus.BAD_REQUEST)
        target, project = req.get('target'), req.get('project')
        if target not in TARGETS or project not in PROJECTS:
            return self.send_json({'error': 'unknown target/project'}, HTTPStatus.BAD_REQUEST)
        return self.send_json(build(target, project))


def main():
    ap = argparse.ArgumentParser(description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)
    ap.add_argument('--port', type=int, default=8732)
    args = ap.parse_args()
    bin_dir, version = find_toolchain()
    print('STM32 Forge server on http://localhost:%d/' % args.port)
    print('toolchain: %s' % (version or 'NOT FOUND (set ARM_GCC_PATH)'))
    ThreadingHTTPServer(('127.0.0.1', args.port), Handler).serve_forever()


if __name__ == '__main__':
    main()

File: server/forge_server.mjs
#!/usr/bin/env node
// Local STM32 Forge API and static server. Requires Node.js and arm-none-eabi-gcc.
import http from 'node:http';
import fs from 'node:fs';
import path from 'node:path';
import os from 'node:os';
import { fileURLToPath } from 'node:url';
import { spawn } from 'node:child_process';

const ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '..');
const PORT = Number(process.env.PORT || 8732);
const TARGETS = {
  h743: { name: 'STM32H743IITx', core: 'Cortex-M7 @ 400 MHz', flash: 2097152, ram: 1048576 },
  bluepill: { name: 'STM32F103C8T6 (Blue Pill)', core: 'Cortex-M3 @ 72 MHz', flash: 65536, ram: 20480 },
};
const PROJECTS = JSON.parse(fs.readFileSync(path.join(ROOT, 'firmware/projects/catalog.json'), 'utf8'))
  .projects.map((project) => project.id);
const EDITABLE = [];
for (const base of ['firmware/projects', 'firmware/targets', 'firmware/third_party', 'engine/platform', 'engine/doomgeneric']) {
  const walk = (dir) => {
    for (const entry of fs.readdirSync(path.join(ROOT, dir), { withFileTypes: true })) {
      const rel = `${dir}/${entry.name}`;
      if (entry.isDirectory()) walk(rel);
      else if (/\.(c|h|s|ld|mk)$/.test(entry.name) || entry.name === 'Makefile') EDITABLE.push(rel);
    }
  };
  walk(base);
}
EDITABLE.push('firmware/Makefile');
EDITABLE.sort();
const editableSet = new Set(EDITABLE);
let building = false;

function toolchainDir() {
  if (process.env.ARM_GCC_PATH) return process.env.ARM_GCC_PATH;
  const onPath = (process.env.PATH || '').split(path.delimiter)
    .find((dir) => fs.existsSync(path.join(dir, 'arm-none-eabi-gcc')));
  if (onPath) return onPath;
  const candidates = [];
  for (const root of [path.join(os.homedir(), '.local/toolchains'), '/opt']) {
    if (!fs.existsSync(root)) continue;
    for (const name of fs.readdirSync(root)) {
      const dir = path.join(root, name, 'bin');
      if (fs.existsSync(path.join(dir, 'arm-none-eabi-gcc'))) candidates.push(dir);
    }
  }
  return candidates.sort().at(-1) || null;
}

function json(res, status, value) {
  const data = Buffer.from(JSON.stringify(value));
  res.writeHead(status, { 'content-type': 'application/json; charset=utf-8', 'content-length': data.length, 'cache-control': 'no-store' });
  res.end(data);
}
function body(req) {
  return new Promise((resolve, reject) => {
    const chunks = [];
    req.on('data', (chunk) => chunks.push(chunk));
    req.on('end', () => resolve(Buffer.concat(chunks)));
    req.on('error', reject);
  });
}
function filePath(rel) {
  const full = path.resolve(ROOT, rel);
  if (!full.startsWith(ROOT + path.sep) || !editableSet.has(rel)) return null;
  return full;
}
function run(command, args, options = {}) {
  return new Promise((resolve) => {
    const child = spawn(command, args, { cwd: ROOT, ...options });
    let stdout = '', stderr = '';
    child.stdout?.on('data', (b) => { stdout += b; });
    child.stderr?.on('data', (b) => { stderr += b; });
    child.on('error', (e) => resolve({ code: 127, stdout, stderr: `${stderr}${e.message}\n` }));
    child.on('close', (code) => resolve({ code: code ?? 1, stdout, stderr }));
  });
}
function parseMemory(log) {
  const rows = [];
  const unit = { B: 1, KB: 1024, MB: 1048576, GB: 1073741824 };
  for (const line of log.split(/\r?\n/)) {
    const m = line.match(/^\s*(\w+):\s+(\d+(?:\.\d+)?)\s*([KMG]?B)\s+(\d+(?:\.\d+)?)\s*([KMG]?B)\s+/);
    if (m && m[1] !== 'Memory') rows.push({ region: m[1], used: Math.round(Number(m[2]) * unit[m[3]]), size: Math.round(Number(m[4]) * unit[m[5]]), overflow: 0 });
  }
  for (const m of log.matchAll(/region [`'](\w+)' overflowed by (\d+) bytes/g)) {
    const r = rows.find((row) => row.region === m[1]);
    if (r) r.overflow = Number(m[2]);
  }
  return rows;
}
async function handleBuild(req, res) {
  if (building) return json(res, 409, { ok: false, log: 'A build is already running.\n', memory: [] });
  let input;
  try { input = JSON.parse((await body(req)).toString('utf8')); }
  catch { return json(res, 400, { error: 'invalid JSON' }); }
  const { target, project } = input;
  if (!Object.hasOwn(TARGETS, target) || !PROJECTS.includes(project)) return json(res, 400, { error: 'unsupported target or project' });
  building = true;
  try {
    const args = ['-C', path.join(ROOT, 'firmware'), `-j${os.cpus().length}`, `TARGET=${target}`, `PROJECT=${project}`];
    const gccPath = toolchainDir();
    const output = path.join(ROOT, 'firmware/build', `${target}-${project}`);
    for (const name of ['firmware.elf', 'firmware.bin', 'firmware.hex']) {
      try { fs.unlinkSync(path.join(output, name)); } catch (e) { if (e.code !== 'ENOENT') throw e; }
    }
    const result = await run('make', args, { env: { ...process.env, ...(gccPath ? { ARM_GCC_PATH: gccPath } : {}) } });
    const log = result.stdout + result.stderr;
    const answer = { ok: result.code === 0, toolchain: '', command: `make ${args.slice(3).join(' ')}`, log, memory: parseMemory(log) };
    if (answer.ok) {
      const binary = fs.readFileSync(path.join(output, 'firmware.bin'));
      answer.binary = binary.toString('base64');
      answer.binarySize = binary.length;
      const nmPath = gccPath ? path.join(gccPath, 'arm-none-eabi-nm') : 'arm-none-eabi-nm';
      const nm = await run(nmPath, [path.join(output, 'firmware.elf')]);
      const symbols = new Map();
      for (const line of nm.stdout.split(/\r?\n/)) {
        const m = line.match(/^([0-9a-fA-F]+)\s+\w\s+(__zone\d_(?:start|end))$/);
        if (m) symbols.set(m[2], parseInt(m[1], 16));
      }
      const banks = [];
      for (let i = 0; symbols.has(`__zone${i}_start`) && symbols.has(`__zone${i}_end`); i++) banks.push(symbols.get(`__zone${i}_end`) - symbols.get(`__zone${i}_start`));
      answer.zoneBanks = banks.length ? banks : null;
      answer.zone = banks.length ? banks.reduce((a, b) => a + b, 0) : null;
    }
    return json(res, 200, answer);
  } finally { building = false; }
}

const MIME = { '.html': 'text/html', '.css': 'text/css', '.js': 'text/javascript', '.json': 'application/json', '.wasm': 'application/wasm', '.wad': 'application/octet-stream', '.bin': 'application/octet-stream', '.png': 'image/png', '.svg': 'image/svg+xml' };
const server = http.createServer(async (req, res) => {
  try {
    const url = new URL(req.url, `http://${req.headers.host || 'localhost'}`);
    if (url.pathname === '/api/status' && req.method === 'GET') {
      const binDir = toolchainDir();
      const gcc = binDir ? path.join(binDir, 'arm-none-eabi-gcc') : 'arm-none-eabi-gcc';
      const v = await run(gcc, ['--version']);
      return json(res, 200, { toolchain: v.code ? null : v.stdout.split(/\r?\n/)[0], toolchainPath: binDir, targets: TARGETS, projects: PROJECTS });
    }
    if (url.pathname === '/api/tree' && req.method === 'GET') return json(res, 200, { files: EDITABLE });
    if (url.pathname === '/api/file') {
      const rel = url.searchParams.get('path') || '';
      const full = filePath(rel);
      if (!full || !fs.existsSync(full)) return json(res, 404, { error: 'not found' });
      if (req.method === 'GET') return json(res, 200, { path: rel, content: fs.readFileSync(full, 'utf8') });
      if (req.method === 'PUT') {
        fs.writeFileSync(full, await body(req));
        return json(res, 200, { ok: true });
      }
      return json(res, 405, { error: 'method not allowed' });
    }
    if (url.pathname === '/api/build' && req.method === 'POST') return await handleBuild(req, res);
    if (req.method !== 'GET' && req.method !== 'HEAD') { res.writeHead(405).end(); return; }
    let rel = decodeURIComponent(url.pathname).replace(/^\/+/, '') || 'index.html';
    let full = path.resolve(ROOT, rel);
    if (!full.startsWith(ROOT + path.sep) || !fs.existsSync(full) || !fs.statSync(full).isFile()) { res.writeHead(404).end('Not found'); return; }
    const stat = fs.statSync(full);
    res.writeHead(200, { 'content-type': `${MIME[path.extname(full)] || 'application/octet-stream'}; charset=utf-8`, 'content-length': stat.size, 'cache-control': 'no-store' });
    if (req.method === 'HEAD') res.end(); else fs.createReadStream(full).pipe(res);
  } catch (e) {
    if (!res.headersSent) json(res, 500, { error: e.message }); else res.destroy(e);
  }
});
server.listen(PORT, '127.0.0.1', () => console.log(`STM32 Forge at http://localhost:${PORT}/`));
